### ISSUE TITLE
Daha kuvvetli kod

### ISSUE THREAD

**Issue by i492095-beep** [i492095-beep/Ismail-Newton#2]
# --- PRO SEVİYE BITCOIN TAHMİN ALGORİTMASI (GPU OPTİMİZASYONLU) ---
# Gerekli kütüphaneler
!pip install tensorflow pandas numpy yfinance scikit-learn matplotlib

import numpy as np
import pandas as pd
import yfinance as yf
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout, Bidirectional
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau

# --- DONANIM KONTROLÜ ---
print("--- Donanım Kontrolü ---")
gpus = tf.config.list_physical_devices('GPU')
if gpus:
    print(f"✅ GPU Tespit Edildi: {gpus[0]}")
    print("RTX 4060 Devrede! Eğitim çok daha hızlı olacak.")
else:
    print("⚠️ GPU Bulunamadı. CPU kullanılacak (Yine de çalışır ama yavaş olur).")

# --- 1. VERİ TOPLAMA VE ÖZELLİK MÜHENDİSLİĞİ (FEATURE ENGINEERING) ---
# Sadece fiyata bakmak yetmez, indikatörleri de ekliyoruz.

data = yf.download('BTC-USD', start='2018-01-01', end='2025-11-17', progress=False)

# Veri setini temizleyelim (Sadece Close ve Volume alalım)
df = data[['Close', 'Volume']].copy()

# --- İndikatörleri Hesapla (Teknik Analiz) ---
# 1. RSI (Göreceli Güç Endeksi)
def compute_rsi(data, window=14):
    delta = data.diff()
    gain = (delta.where(delta > 0, 0)).rolling(window=window).mean()
    loss = (-delta.where(delta < 0, 0)).rolling(window=window).mean()
    rs = gain / loss
    return 100 - (100 / (1 + rs))

df['RSI'] = compute_rsi(df['Close'])

# 2. MACD (Hareketli Ortalama Yakınsama Iraksama)
df['EMA12'] = df['Close'].ewm(span=12, adjust=False).mean()
df['EMA26'] = df['Close'].ewm(span=26, adjust=False).mean()
df['MACD'] = df['EMA12'] - df['EMA26']

# 3. SMA (Basit Hareketli Ortalama - 50 Günlük)
df['SMA50'] = df['Close'].rolling(window=50).mean()

# NaN (Boş) verileri temizle (İndikatörler ilk günlerde hesaplanamaz)
df = df.dropna()

print(f"İşlenen Veri Boyutu: {df.shape}")
print("Kullanılan Özellikler: Close, Volume, RSI, MACD, SMA50")

# --- 2. VERİ ÖLÇEKLENDİRME (SCALING) ---
# İki ayrı scaler kullanacağız. Biri girişler (tüm özellikler) için, biri çıkış (sadece fiyat) için.
# Bu, sonradan tahmini geri çevirmeyi (inverse transform) çok kolaylaştırır.

# Hedef (Tahmin edilecek şey): Sadece 'Close' fiyatı
target_scaler = MinMaxScaler(feature_range=(0, 1))
scaled_target = target_scaler.fit_transform(df[['Close']])

# Girişler (Modelin bakacağı her şey): Close, Volume, RSI, MACD, SMA50
feature_scaler = MinMaxScaler(feature_range=(0, 1))
scaled_features = feature_scaler.fit_transform(df[['Close', 'Volume', 'RSI', 'MACD', 'SMA50']])

# --- 3. PENCERELEME (WINDOWING) ---
# Donanımın güçlü, 60 gün yerine 90 güne bakabiliriz. Daha fazla bağlam = daha iyi tahmin.
window_size = 90

X = []
y = []

for i in range(window_size, len(scaled_features)):
    X.append(scaled_features[i-window_size:i]) # Geçmiş 90 gün (tüm özellikler)
    y.append(scaled_target[i])              # Bugünün kapanış fiyatı

X, y = np.array(X), np.array(y)

# Eğitim ve Test ayırma (%80 - %20)
split_idx = int(len(X) * 0.8)
X_train, X_test = X[:split_idx], X[split_idx:]
y_train, y_test = y[:split_idx], y[split_idx:]

print(f"Eğitim Verisi: {X_train.shape}") # (Örnek Sayısı, 90 Gün, 5 Özellik)

# --- 4. GELİŞMİŞ MODEL MİMARİSİ (RTX 4060 POWER) ---
model = Sequential()

# 1. Katman: Bidirectional LSTM (Geçmişi ve Geleceği Görür)
# 128 Nöron (RAM'in var, kullan!)
model.add(Bidirectional(LSTM(units=128, return_sequences=True), input_shape=(X_train.shape[1], X_train.shape[2])))
model.add(Dropout(0.3)) # Ezberlemeyi önle

# 2. Katman: Stacked LSTM
model.add(LSTM(units=128, return_sequences=True))
model.add(Dropout(0.3))

# 3. Katman: Stacked LSTM (Derinleşiyoruz)
model.add(LSTM(units=64, return_sequences=False))
model.add(Dropout(0.2))

# 4. Yoğun Katman (Feature Extraction)
model.add(Dense(units=50, activation='relu'))

# 5. Çıkış Katmanı
model.add(Dense(units=1)) # Fiyat Tahmini

# Derleme
model.compile(optimizer='adam', loss='mean_squared_error')

# --- 5. EĞİTİM STRATEJİSİ (CALLBACKS) ---
# Eğitim sırasında model iyileşmezse durdur veya öğrenme hızını düşür.

callbacks = [
    EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True, verbose=1),
    ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=5, min_lr=0.00001, verbose=1)
]

# Eğitimi Başlat (RTX 4060 burada fan açabilir :) )
# Batch size 64 veya 128 yapabiliriz (GPU belleğin yeterli)
history = model.fit(X_train, y_train,
                    epochs=100, # Erken durdurma var, 100 diyebiliriz
                    batch_size=64,
                    validation_data=(X_test, y_test),
                    callbacks=callbacks,
                    verbose=1)

# --- 6. TAHMİN VE GÖRSELLEŞTİRME ---
train_predict = model.predict(X_train)
test_predict = model.predict(X_test)

# Ölçekleri geri çevir (Gerçek $ değerleri)
train_predict = target_scaler.inverse_transform(train_predict)
y_train_real = target_scaler.inverse_transform(y_train)
test_predict = target_scaler.inverse_transform(test_predict)
y_test_real = target_scaler.inverse_transform(y_test)

# Grafik
plt.figure(figsize=(16, 8))
plt.title('Gelişmiş Bitcoin Fiyat Tahmini (Çok Değişkenli LSTM)', fontsize=18)
plt.xlabel('Zaman', fontsize=14)
plt.ylabel('Fiyat (USD)', fontsize=14)
plt.plot(y_test_real, label='Gerçek Fiyat', color='#1f77b4', linewidth=2)
plt.plot(test_predict, label='Model Tahmini', color='#ff7f0e', linestyle='--', linewidth=2)
plt.legend(fontsize=14)
plt.grid(True, alpha=0.3)
plt.show()

# --- 7. YARINI TAHMİN ET (GELECEK) - DÜZELTİLMİŞ VERSİYON ---
# En son 90 günlük veriyi (Tüm özellikleri içeren) al
last_sequence = scaled_features[-window_size:]
last_sequence = np.reshape(last_sequence, (1, window_size, 5)) # 5 Özellik var!

future_pred_scaled = model.predict(last_sequence)
future_pred = target_scaler.inverse_transform(future_pred_scaled)

# HATA ÇÖZÜMÜ BURADA: .item() kullanarak veriyi saf sayıya çeviriyoruz
last_real_price = df['Close'].iloc[-1]
if isinstance(last_real_price, pd.Series):
    last_real_price = last_real_price.item()

print("\n" + "="*40)
print(f"🔍 BİLİNEN SON FİYAT:  {float(last_real_price):.2f} $")
print(f"🤖 MODELİN YARIN TAHMİNİ: {float(future_pred[0][0]):.2f} $")
print("="*40)

if future_pred > last_real_price:
    print("🚀 Sinyal: YÜKSELİŞ Bekleniyor")
else:
    print("🔻 Sinyal: DÜŞÜŞ Bekleniyor")
